gl_state_cache

juj opened the pull request:
Since WebGL calls can be quite slow, implement an opt in state cache for commonly set GL state variables. This can improve performance in applications which perform some redundant GL calls that have no effect.

kripken commented:
Before I read the details here, if this is something applications can fix themselves, shouldn't we encourage them to do so instead of working around their inefficiency (especially when it adds complexity to us, which at first glance at this PR seems to be the case, but I could be wrong)?

juj commented:
Updated this to also not inject a compiled file, and added a test.

We indeed do want people to optimize their codebases primarily. I'm writing a doc page to add to Emscripten for optimizing WebGL performance, which will help there. However I've written so many GL state caches while optimizing various GL codebases for WebGL, that having a cache directly in Emscripten for the most common GL state variables makes it easier, and it is then simple for people to quickly toggle a linker flag to test what kind of effect it will have.

The idea is not to fully cache all GL variables, but a few common ones that we see causing a noticeable impact in different cases. (Although with --proxy-to-worker GL state caching, it may become the case that we'll want to proxy way much more data like the proxied GL lib now does, except in C/C++ side for performance and so the state is shared between pthreads)

kripken commented on src/library_gl.js:
```diff
@@ -1296,6 +1296,27 @@ var LibraryGL = {
+#if GL_STATE_CACHE
+    if (target == 0x0de1 /* GL_TEXTURE_2D */) {
+      var e = GL.textures[GL.currentlyBoundTexture[GL.currentlyActiveTexture]];
+      if (typeof e !== 'undefined') {
```
can do `e !== undefined`

kripken replied:
(and other places below)

kripken commented on src/library_gl.js:
```diff
@@ -7492,7 +7555,7 @@ var LibraryGL = {
 
 // Simple pass-through functions. Starred ones have return values. [X] ones have X in the C name but not in the JS name
 var glFuncs = [[0, 'finish flush'],
- [1, 'clearDepth clearDepth[f] depthFunc enable disable frontFace cullFace clear lineWidth clearStencil stencilMask checkFramebufferStatus* generateMipmap activeTexture blendEquation isEnabled*'],
```
what is this change?

juj replied:
This one removes `glActiveTexture` from getting autogenerated, since in the `GL_STATE_CACHE` enabled case, we need to remember what the active texture was (at https://github.com/kripken/emscripten/pull/4898/files#diff-78a51faf7f8c4f62113aae3d5d4ae8eeR1384)

kripken replied:
I see, thanks.

kripken commented on tests/webgl_state_cache.js:
```diff
@@ -0,0 +1,20 @@
+mergeInto(LibraryManager.library, {
+	activeAttribs: 0,
```
indentation seems excessive

juj replied:
Yeah, this one had tabs as indents, converting to spaces.

kripken commented on src/library_gl.js:
```diff
@@ -3817,6 +3856,30 @@ var LibraryGL = {
 #endif
   },
 
+#if GL_STATE_CACHE
```
is this just for testing? i don't understand this

juj replied:
These one exist to route the default names for these functions to other names, which can then be referenced from `gl_cache.c`, which contain the injected versions. So applications first call the GL functions in `gl_cache.c`, which then call these renamed functions.

kripken replied:
I see. Perhaps add a comment explaining this?

Another question, does this not interfere with the glProcAddress stuff, for which we already have another alias for each GL call (I think it was also `emscripten_` or similar)? Or is that why this conflicts with gl emulation?

juj commented:
Thanks @dos1, good find. This PR never landed, and unfortunately I don't currently have a chance to work on this. Posted PR #7959 to remove the stale docs. I'll close this PR as old.
